server/src/main/java/org/elasticsearch/index/store/AsyncDirectIOIndexInput.java

```diff
@@ -0,0 +1,539 @@
+                }
+            }
+        }
+    }
+
+    @Override
+    public void readFloats(float[] dst, int offset, int len) throws IOException {
+        int remainingDst = len;
+        while (remainingDst > 0) {
+            int cnt = Math.min(buffer.remaining() / Float.BYTES, remainingDst);
+            buffer.asFloatBuffer().get(dst, offset + len - remainingDst, cnt);
+            buffer.position(buffer.position() + Float.BYTES * cnt);
+            remainingDst -= cnt;
+            if (remainingDst > 0) {
+                if (buffer.hasRemaining()) {
+                    dst[offset + len - remainingDst] = Float.intBitsToFloat(readInt());
+                    --remainingDst;
+                } else {
+                    refill(remainingDst * Float.BYTES);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void readLongs(long[] dst, int offset, int len) throws IOException {
+        int remainingDst = len;
+        while (remainingDst > 0) {
+            int cnt = Math.min(buffer.remaining() / Long.BYTES, remainingDst);
+            buffer.asLongBuffer().get(dst, offset + len - remainingDst, cnt);
+            buffer.position(buffer.position() + Long.BYTES * cnt);
+            remainingDst -= cnt;
+            if (remainingDst > 0) {
+                if (buffer.hasRemaining()) {
+                    dst[offset + len - remainingDst] = readLong();
+                    --remainingDst;
+                } else {
+                    refill(remainingDst * Long.BYTES);
+                }
+            }
+        }
+    }
+
+    @Override
+    public AsyncDirectIOIndexInput clone() {
+        try {
+            var clone = new AsyncDirectIOIndexInput("clone:" + this, this, offset, length);
+            // TODO figure out how to make this async
+            clone.seekInternal(getFilePointer());
+            return clone;
+        } catch (IOException ioe) {
+            throw new UncheckedIOException(ioe);
+        }
+    }
+
+    @Override
+    public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+        if ((length | offset) < 0 || length > this.length - offset) {
+            throw new IllegalArgumentException("slice() " + sliceDescription + " out of bounds: " + this);
+        }
+        var slice = new AsyncDirectIOIndexInput(sliceDescription, this, this.offset + offset, length);
+        // TODO figure out how to make this async
+        slice.seekInternal(0L);
+        return slice;
+    }
+
+    /**
+     * A simple prefetcher that uses virtual threads to prefetch data into direct byte buffers.
+     */
+    private static class DirectIOPrefetcher implements Closeable {
+        private final int maxConcurrentPrefetches;
+        private final int maxTotalPrefetches;
+        private final FileChannel channel;
+        private final int blockSize;
+        private final long[] prefetchPos;
+        private final List<Future<ByteBuffer>> prefetchThreads;
```

Comment:
Is the only reason this is a List so the generics all work? Is there a way it could be a simple array?

Reply:
I tried the array, but we need `Future<Object>` for the underlying logic to work with the executor.

Reply:
++